include/uk: Move compiler definitions from `essentials.h` to `compiler.h`

rares-miculescu opened the pull request:
### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Compiler definitions from `include/uk/essentials.h` (including `container_of`) are moved to `include/uk/compiler.h`.

michpappas reviewed (changes requested):
Hi @rares-miculescu to remove the `Draft` status I would recommend the following changes:

1. Rebase your tree to staging and remove unnecessary commits from other PRs. I am referring to the one that you add and later revert. Generally speaking if there are dependencies you can include in the tree but clearly mark them as such (similarly to what @mogasergiu does in #912) and leave a note in your PR's cover letter. If they are not dependencies they simply don't belong there.
2. Squash your remaining three commits into one. Moving definitions from `essentials.h` to `compiler.h` consists of a single logic operation, so it can be done in a single commit.
3. Provide a proper commit message with a short description and a `Signed-off-by`. You can see other commit messages for reference.

michpappas commented on include/uk/essentials.h:
```diff
@@ -274,46 +274,6 @@ extern "C" {
 #define DEQUALIFY(t, a) ((t)(__uptr)(const volatile void *)(a))
 #endif
 
-/* The following code was taken from FreeBSD: sys/sys/cdefs.h */
```
Now that this is moved to a different file, it may required that we remove the `Copyright of University of California` from the header of `essentials.h`. Yet it is unclear to me what is the limit of the imported code, but I suspect the remaining part starting from `UKCTASSERT` is original work. @skuenzer perhaps remembers?

michpappas commented on include/uk/compiler.h:
```diff
@@ -0,0 +1,95 @@
+#define __UK_COMPILER_H__
+
+#include <uk/config.h>
+
```
You should add `#include <asm/compiler.h>` here, as that imports `compile.h` of the architecture we compile for (`arch/$ARCH/include/uk/asm/compiler.h`).

michpappas commented on include/uk/essentials.h:
```diff
@@ -274,46 +274,6 @@ extern "C" {
-#endif
-#endif /* !__containerof */
-
 /* Computes the offsset and size of a field within a struct and checks
```
I think this belongs to `compiler.h` too, being a derivative of `__offsetof`.

michpappas reviewed (changes requested):
Hi @rares-miculescu thanks for updating this. Just a few minor comments.

StefanJum commented on include/uk/compiler.h:
```diff
@@ -0,0 +1,77 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2017, NEC Europe Ltd., NEC Corporation. All rights reserved.
+ * Copyright (c) 1991, 1993
+ *	The Regents of the University of California.  All rights reserved.
+ * Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
I would add this in chronological order.

StefanJum commented on include/uk/essentials.h:
```diff
@@ -1,38 +1,11 @@
- *
- *
- * Copyright (c) 2017, NEC Europe Ltd., NEC Corporation. All rights reserved.
+/* Copyright (c) 2017, NEC Europe Ltd., NEC Corporation. All rights reserved.
  * Copyright (c) 1991, 1993
  *	The Regents of the University of California.  All rights reserved.
```
Same here, I would order the copyrights.

StefanJum reviewed (commented):
Hi @rares-miculescu, few comments below.

Besides those, please wrap your commit message to about 50 characters (see [this guide](https://cbea.ms/git-commit/)).

Also it might make sense to have the comments `doxygen-style` (an example [here](https://github.com/unikraft/unikraft/blob/staging/lib/9pfs/9pfs.h)), since we already modify the headers. We could also do this in a separate pr and keep this just for moving stuff (@razvand  @michpappas what do you think?).

Commit:
include/uk: Move compiler definitions to own header

Move compiler definitions from `include/uk/essentials.h`
to `include/uk/compiler.h`, for modularity enhancement
and easier maintenance.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

michpappas commented on include/uk/essentials.h:
```diff
@@ -54,105 +27,7 @@
-#else
-/* TO BE DEFINED */
-#endif /* __GNUC__ */
+#include <uk/compiler.h>
```
This include should be moved up together with `uk/config.h`, and `compile.h` should protect itself with C++ guards instead.

michpappas commented on include/uk/compiler.h:
```diff
@@ -0,0 +1,178 @@
+#define __constructor_prio(lvl) __attribute__ ((constructor (lvl)))
+#endif
+
+#ifdef CONFIG_HAVE_SCHED
```
These are not compiler related, it's okay to leave them in `essentials.h`.

michpappas reviewed (changes requested):
Hi @rares-miculescu thanks for updating this. Only a couple of comments from my side.

michpappas commented:
Also @rares-miculescu when you update this please remove the `Draft` state.

michpappas commented on include/uk/essentials.h:
```diff
@@ -1,43 +1,20 @@
 
 #include <uk/config.h>
 
+#ifndef __cplusplus
```
What I meant in my previous comment is that instead of doing this, you can include `uk/compiler.h` unconditionally, and add C++ guards inside `compiler.h`.

michpappas reviewed (changes requested):
@rares-miculescu thanks for the updates. Besides the comment below, you can also consider improving a bit the commit message to something like:
```
include/uk: Move compiler definitions to own header
    
Move compiler definitions from `include/uk/essentials.h`
to `include/uk/compiler.h`.
```

michpappas commented on include/uk/compiler.h:
```diff
@@ -0,0 +1,176 @@
+#include <uk/asm/compiler.h>
+
+#ifdef __cplusplus
+extern "C" {
```
Where does this brace close?

michpappas commented on include/uk/compiler.h:
```diff
@@ -0,0 +1,195 @@
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef _UK_COMPILER_H_
```
One last change: Please change these to `__UK_COMPILER_H__` here and at the comment at the end of the file. Thanks!

michpappas reviewed (approved):
LGTM, thanks @rares-miculescu

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
